src/tlo/analysis/life_expectancy.py

```diff
@@ -0,0 +1,305 @@
+    # add adjusted data back into dataframe
+    num_males.iloc[0] = num_males_adj.values
+    # change first row index
+    num_males.index.values[0] = '1-4'
+    # add in infant numbers to main dataframe
+    t1 = pd.concat([num_m0, num_males])
+    # set correct index
+    t1.index.values[0] = '0'
+
+    # repeat for females
+    num_females_adj = num_females.iloc[0:1, :] - num_f0.iloc[0:1, :].values
+    num_females.iloc[0] = num_females_adj.values
+    num_females.index.values[0] = '1-4'
+    t2 = pd.concat([num_f0, num_females])
+    t2.index.values[0] = '0'
+
+    # combine pop aged 90+
+    # Sum the last three rows column-wise
+    last_three_rows_sum = t1.iloc[-3:].sum()
+    t1.loc['90+'] = last_three_rows_sum
+
+    last_three_rows_sum = t2.iloc[-3:].sum()
+    t2.loc['90+'] = last_three_rows_sum
+
+    # remove unneeded columns
+    indices_to_remove = ['90-94', '95-99', '100+']
+    t1 = t1.drop(indices_to_remove, axis=0)
+    t2 = t2.drop(indices_to_remove, axis=0)
+
+    # output is 0, 1-4, 5-9, 10-14 and index by male/female
+    # join male and female dataframe and add index for sex
+    t3 = pd.concat([t1, t2], ignore_index=True)
+
+    # Create a MultiIndex for rows using age group and 'Male' and 'Female'
+    multi_index_rows_male = pd.MultiIndex.from_product([t1.index, ['M']], names=['age_group', 'sex'])
+    multi_index_rows_female = pd.MultiIndex.from_product([t1.index, ['F']], names=['age_group', 'sex'])
+    t3.index = multi_index_rows_male.append(multi_index_rows_female)
+
+    return t3
+
+
+# %% GENERATE LIFE EXPECTANCY ESTIMATES
+
+def _estimate_life_expectancy(person_years_at_risk, number_of_deaths_in_interval):
+    """
+    for a single run, estimate life expectancy for males and females
+    return: pd.Series
+    """
+
+    estimated_life_expectancy = pd.Series(dtype=float)
+
+    # first age-group is 0, then 1-4, 5-9, 10-14 etc. 22 categories in total
+    level_0_values = person_years_at_risk.index.get_level_values('age_group').unique()
+    interval_width = pd.Series([1] + [4] + [5] * 18, index=level_0_values)
+
+    fraction_of_last_age_survived = 0.5
+    number_age_groups = 20
+
+    # separate male and female data
+    for sex in ['M', 'F']:
+        person_years_by_sex = person_years_at_risk.xs(key=sex, level='sex')
+        number_of_deaths_by_sex = number_of_deaths_in_interval.xs(key=sex, level='sex')
+
+        death_rate_in_interval = number_of_deaths_by_sex / person_years_by_sex
+        # if no deaths or person-years, convert nan to 0
+        death_rate_in_interval = death_rate_in_interval.fillna(0)
+        # if no deaths in age 90+, set death rate equal to value in age 85-89
+        if death_rate_in_interval[number_age_groups - 1] == 0:
+            death_rate_in_interval[number_age_groups - 1] = death_rate_in_interval[number_age_groups - 2]
+
+        # Calculate the probability of dying in the interval
+        condition = number_of_deaths_by_sex > (
```

Comment:
what does condition refer to? (suggest to add comments)

Reply:
I've added this explanation now: 
        # Calculate the probability of dying in the interval
        # condition checks whether the observed number deaths is significantly higher than would be expected
        # based on population years at risk and survival fraction
        # if true, suggests very high mortality rates and returns value 1